upup/pkg/fi/cloudup/gce/tpm/gcetpmverifier/tpmverifier.go

```diff
@@ -0,0 +1,241 @@
+	"k8s.io/kops/pkg/apis/nodeup"
+	"k8s.io/kops/pkg/bootstrap"
+	"k8s.io/kops/pkg/nodeidentity/gce"
+	"k8s.io/kops/upup/pkg/fi"
+	tpm "k8s.io/kops/upup/pkg/fi/cloudup/gce/tpm"
+)
+
+type tpmVerifier struct {
+	opt tpm.TPMVerifierOptions
+
+	computeClient *compute.Service
+}
+
+// NewTPMVerifier constructs a new TPM verifier for GCE.
+func NewTPMVerifier(opt *tpm.TPMVerifierOptions) (bootstrap.Verifier, error) {
+	ctx := context.Background()
+
+	computeClient, err := compute.NewService(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error building compute API client: %w", err)
+	}
+
+	return &tpmVerifier{
+		opt:           *opt,
+		computeClient: computeClient,
+	}, nil
+}
+
+var _ bootstrap.Verifier = &tpmVerifier{}
+
+func (v *tpmVerifier) Verify(ctx context.Context, token string, body []byte) (*bootstrap.VerifyResult, error) {
+	bootstrapRequest := &nodeup.BootstrapRequest{}
+	if err := json.Unmarshal(body, bootstrapRequest); err != nil {
+		return nil, fmt.Errorf("failed to decode request: %w", err)
+	}
+
+	claim := bootstrapRequest.Claim
+	if claim == nil {
+		return nil, fmt.Errorf("claim is required")
+	}
+
+	node := claim.Node
+	if claim.Node == nil {
+		return nil, fmt.Errorf("node is required")
+	}
+
+	if node.GCPProjectID == "" {
+		return nil, fmt.Errorf("gcpProjectID is required")
+	}
+	if node.Zone == "" {
+		return nil, fmt.Errorf("zone is required")
+	}
+	if node.Instance == "" {
+		return nil, fmt.Errorf("instance is required")
+	}
+
+	if node.GCPProjectID != v.opt.ProjectID {
+		return nil, fmt.Errorf("projectID does not match expected: got %q, want %q", node.GCPProjectID, v.opt.ProjectID)
+	}
+
+	instance, err := v.computeClient.Instances.Get(node.GCPProjectID, node.Zone, node.Instance).Do()
+	if err != nil {
+		if isNotFound(err) {
+			return nil, fmt.Errorf("unable to find instance in compute API: %w", err)
+		}
+		return nil, fmt.Errorf("error fetching instance from compute API: %w", err)
+	}
+
+	if !strings.HasPrefix(lastComponent(instance.Zone), v.opt.Region+"-") {
+		return nil, fmt.Errorf("instance was in zone %q, expected region %q", instance.Zone, v.opt.Region)
+	}
+
+	// Verify node is in our cluster
+	clusterName := ""
+	instanceGroupName := ""
+	for _, item := range instance.Metadata.Items {
+		switch item.Key {
+		case gce.MetadataKeyInstanceGroupName:
+			instanceGroupName = fi.StringValue(item.Value)
+		case "cluster-name":
+			clusterName = fi.StringValue(item.Value)
+		}
+	}
+
+	if clusterName == "" {
+		return nil, fmt.Errorf("could not determine cluster for instance %s", instance.SelfLink)
+	}
+
+	if clusterName != v.opt.ClusterName {
+		return nil, fmt.Errorf("clusterName does not match expected: got %q, want %q", clusterName, v.opt.ClusterName)
+	}
+	if instanceGroupName == "" {
+		return nil, fmt.Errorf("could not determine instance group for instance %s", instance.SelfLink)
+	}
+
+	// TODO can we include the attestation certificate (signed by GCE) in the claim, rather than getting it from the API?
```

Comment:
If the attestation certificate also contains the zone, instance metadata and network interfaces you might be able to avoid using the API altogether. Otherwise, I don't see the value in merely reducing the number of calls.

Getting it from the API validates that the instance still exists.

Reply:
Downgraded from a TODO to a Note.  It's a cute trick, at least, and might be the default way for other TPM implementations.